### PR TITLE
Additions to distributed ops, target ops, and new rollout APIs.

### PR DESCRIPTION
**Pull request by wenkesj**

### CODE REVIEW

File: pyoneer/rl/transitions/buffer_ops.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        def gather_trajectories(buffer):
+            return tf.gather(buffer, trajectory_slots)
+
+        return tf.nest.map_structure(gather_trajectories, self._trajectories)
+
+
+class NstepLRUBuffer(NstepBuffer):
+
+    def write(self, episode_ids, n_step_ids, trajectories):
+        """Write trajectoies into the buffer.
```
Comment (djsaunde) [inline pyoneer/rl/transitions/buffer_ops.py]: `trajectoies` -> `trajectories`

File: pyoneer/rl/transitions/buffer_ops.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    @property
+    def max_size(self):
+        return self._max_size
+
+    @property
+    def non_empty(self):
+        return tf.math.count_nonzero(
+            tf.not_equal(self._trajectory_slots_to_ids, -1))
+
+    def _write(self, trajectory_slots, trajectory_ids, trajectory_pos,
```
Comment (djsaunde) [inline pyoneer/rl/transitions/buffer_ops.py]: Why is this a private method? Is this class not meant to be instantiated?

File: pyoneer/rl/targets/reward_ops.py [hunk 3]
```diff
@@ -0,0 +1,668 @@
...
+        rewards: tensor of shape [Batch x Time], [Batch x Time x ...]
+        bootstrap_value: The last value [t+n]. A tensor of shape [],
+            [...], [Batch], [Batch, ...], [Batch x Time], [Batch x Time x ...].
+        discounts: tensor of shape [], [...], [Batch], [Batch, ...],
+            [Batch x Time], [Batch x Time x ...].
+        weights: tensor of shape [], [...], [Batch], [Batch, ...],
+            [Batch x Time], [Batch x Time x ...].
+        time_major: flag if tensors are time_major.
+            Batch and Time are transposed in this doc.
+        back_prop: allow back_prop through the calculation.
```
Comment (djsaunde) [inline pyoneer/rl/targets/reward_ops.py:29]: OOC, in what circumstances would we want to backprop through the return calculation?

File: pyoneer/rl/targets/reward_ops.py [hunk 4]
```diff
@@ -0,0 +1,668 @@
...
+    Args:
+        rewards: tensor of shape [Batch x Time], [Batch x Time x ...]
+        bootstrap_value: The last value [t+n]. A tensor of shape [],
+            [...], [Batch], [Batch, ...], [Batch x Time], [Batch x Time x ...].
+        discounts: tensor of shape [], [...], [Batch], [Batch, ...],
+            [Batch x Time], [Batch x Time x ...].
+        weights: tensor of shape [], [...], [Batch], [Batch, ...],
+            [Batch x Time], [Batch x Time x ...].
+        time_major: flag if tensors are time_major.
+            Batch and Time are transposed in this doc.
```
Comment (djsaunde) [inline pyoneer/rl/targets/reward_ops.py:28]: What does this mean?

File: pyoneer/rl/targets/reward_ops.py [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+    Returns:
+        tensor of shape [Batch x Time x ...] or [Time x Batch x ...]
+    """
+    with tf.name_scope(name or 'DiscountedReturns'):
+        batch_axis = int(time_major)
+
+        # Check rewards shape.
+        rewards = tf.convert_to_tensor(rewards)
+        rewards_rank = rewards.shape.rank
+        assert rewards_rank >= 2, 'rewards must be atleast rank 2.'
```
Comment (djsaunde) [inline pyoneer/rl/targets/reward_ops.py]: nit: `atleast` -> `at least`

File: pyoneer/rl/targets/reward_ops.py [hunk 6]
```diff
@@ -0,0 +1,668 @@
...
+        rewards_shape = tf.shape(rewards)
+
+        # Compute the batch shape of rewards.
+        if time_major:
+            reward_shape_no_time = rewards_shape[1:]
+        else:
+            reward_shape_no_time = rewards_shape[:1]
```
Comment (djsaunde) [inline pyoneer/rl/targets/reward_ops.py:48]: What if rank of rewards is > 2?

File: pyoneer/rl/targets/reward_ops.py [hunk 7]
```diff
@@ -0,0 +1,668 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+
+@tf.function
+def discounted_returns(rewards,
+                       bootstrap_value=None,
+                       discounts=0.99,
```
Comment (djsaunde) [inline pyoneer/rl/targets/reward_ops.py]: OOC, in what circumstances would we want to use different discount factors?

File: pyoneer/rl/targets/reward_ops.py [hunk 8]
```diff
@@ -0,0 +1,668 @@
...
+            else:
+                bootstrap_value = tf.zeros_like(values[:, -1])
+        else:
+            bootstrap_value = tf.convert_to_tensor(bootstrap_value)
+
+        bootstrap_value_t = tf.expand_dims(bootstrap_value, axis=time_axis)
+
+        # Shift values to tp1.
+        if time_major:
+            values_tp1 = values[1:, :]
```
Comment (djsaunde) [inline pyoneer/rl/targets/reward_ops.py:278]: Could possibly use a more descriptive name than `values_tp1`; not sure what this mean (values from time plus one?)

File: pyoneer/rl/targets/reward_ops.py [hunk 9]
```diff
@@ -0,0 +1,668 @@
...
+def lambda_returns(rewards,
+                   values,
+                   bootstrap_value=None,
+                   discounts=.99,
+                   lambdas=.975,
+                   weights=1.,
+                   time_major=False,
+                   back_prop=False,
+                   name=None):
+    """Computes lambda-returns.
```
Comment (djsaunde) [inline pyoneer/rl/targets/reward_ops.py:209]: What are lambda returns? Doc maybe should specify.

File: pyoneer/rl/targets/reward_ops.py [hunk 10]
```diff
@@ -0,0 +1,668 @@
...
+
+@tf.function
+def discounted_returns(rewards,
+                       bootstrap_value=None,
+                       discounts=0.99,
+                       weights=1.0,
+                       time_major=False,
+                       back_prop=False,
+                       name=None):
+    """Compute discounted returns.
```
Comment (djsaunde) [inline pyoneer/rl/targets/reward_ops.py:17]: What are discounted returns? Doc maybe should specify.

File: pyoneer/rl/targets/reward_ops.py [hunk 11]
```diff
@@ -0,0 +1,668 @@
...
+                   weights=1.,
+                   time_major=False,
+                   back_prop=False,
+                   name=None):
+    """Computes lambda-returns.
+
+    For example, this can be used to compute TD(lambda):
+
+        ```
+        td_lambda = lambda_returns(r(t), V(t+1)) - V(t)
```
Comment (djsaunde) [inline pyoneer/rl/targets/reward_ops.py:214]: Why not just compute TD(lambda) instead of "lambda returns"?

File: pyoneer/rl/targets/reward_ops.py [hunk 12]
```diff
@@ -0,0 +1,668 @@
...
+        if time_major:
+            values_tp1 = values[1:, :]
+        else:
+            values_tp1 = values[:, 1:]
+
+        # Pad the values with the last value.
+        values_tp1 = tf.concat([values_tp1, bootstrap_value_t],
+                               axis=time_axis)
+
+        delta = rewards + discounts * values_tp1 * (1. - lambdas)
```
Comment (djsaunde) [inline pyoneer/rl/targets/reward_ops.py]: Why is this multiplied by `1 - lambdas` instead of by `lambdas`?

File: pyoneer/rl/targets/reward_ops.py [hunk 13]
```diff
@@ -0,0 +1,668 @@
...
+                    values,
+                    log_probs,
+                    log_probs_old,
+                    bootstrap_value=None,
+                    discounts=.99,
+                    weights=1.,
+                    time_major=False,
+                    back_prop=False,
+                    name=None):
+    """Computes v-trace returns.
```
Comment (djsaunde) [inline pyoneer/rl/targets/reward_ops.py:337]: What are v-trace returns? Similarly to above. Maybe pointers to papers would be good here, too.

File: pyoneer/rl/rollouts/unroll_ops.py [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+from __future__ import print_function
+
+import collections
+
+import tensorflow as tf
+
+
+class Unroll(collections.namedtuple(
+        'Unroll', ['outputs', 'finalizer'])):
+    """Creates a new Unroll.
```
Comment (djsaunde) [inline pyoneer/rl/rollouts/unroll_ops.py:12]: What's an unroll?

File: pyoneer/losses/losses_impl.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    Args:
+        losses: Tensor of losses.
+        sample_weight: Tensor of sample weight(s).
+
+    Returns:
+        Weighted loss.
+    """
+    losses = losses * sample_weight
+    total_loss = tf.reduce_sum(losses)
+    num_present = tf.reduce_sum(tf.ones_like(losses) * sample_weight)
```
Comment (djsaunde) [inline pyoneer/losses/losses_impl.py:21]: Why not just `tf.reduce_sum(sample_weight)`?

**Comment by djsaunde** [comment]
Can you also run the black formatter?

**Comment by djsaunde** [comment]
Oh, yeah, and it looks like there some merge conflicts that need resolution.
